Simbody/src/RigidBodyNodeSpec.cpp

```diff
@@ -552,6 +552,84 @@ RigidBodyNodeSpec<dof, noR_FM, noX_MB, noR_PF>::multiplyByMInvPass2Outward(
 }
 
 
+// Base to tip recursion for multiplyBySqrtMInv: 
+// temp allA_GB does not need to be initialized before
+// beginning the iteration.
+template<int dof, bool noR_FM, bool noX_MB, bool noR_PF> void
+RigidBodyNodeSpec<dof, noR_FM, noX_MB, noR_PF>::multiplyBySqrtMInvPass2Outward(
+    const SBInstanceCache&                  ic,
+    const SBTreePositionCache&              pc,
+    const SBArticulatedBodyInertiaCache&    abc,
+    const SBDynamicsCache&                  dc,
+    const Real*                             allEpsilon,
+    SpatialVec*                             allA_GB,
+    Real*                                   allUDot) const
+{
+    const Vec<dof>& eps  = fromU(allEpsilon);
+    SpatialVec&     A_GB = allA_GB[nodeNum];
+    Vec<dof>&       udot = toU(allUDot); // pull out this node's udot
+
+    const bool isPrescribed = isUDotKnown(ic);
+    const HType&        H   = getH(pc);
+    const PhiMatrix&    phi = getPhi(pc);
+    const Mat<dof,dof>& DI  = getDI(abc);
+
+    const HType&              G = getG(abc);
+    SpatialMat tauBar = G*~H;// 11*dof^2 flops 
+    SpatialMat psi = phi*tauBar; // ~100 flops 
+
+    int i, j, k;
+
+    Mat<dof,dof> sqrtDI(1);
+    // DI Cholesky Decomposition: 
+    if(dof>1){
+      double s1, s2;
+      Mat<dof,dof> L(0);
+      Mat<dof,dof> De(0);
+
+      for(j=0; j<dof; j++){
+        L(j,j) = 1.0;
+        // Diagonal
+        s1 = 0.0;
+        for(k=0; k<dof; k++){
+          s1 += (L(j,k) * L(j,k)) * De(k,k);
+        }
+        De(j,j) = DI(j,j) - s1;
+        // Off diagonal
+        for(i=(j+1); i<dof; i++){
+          s2 = 0.0;
+          for(k=0; k<j; k++){
+            s2 += L(i,k) * L(j,k) * De(k,k);
+          }
+          L(i,j) = (1 / De(j,j)) * (DI(i,j) - s2);
+        }
+      }
+
+      // Square root the diagonal
+      for(j=0; j<dof; j++){De(j,j) = sqrt(De(j,j));}
+
+      // Get LsqrtDe
+      sqrtDI = L *  De ;
+    }
+    else{
+      sqrtDI(0,0) = sqrt(DI(0,0));
+    }
+    // END Cholesky Decomposition
+
+    // Shift parent's acceleration outward (Ground==0). 12 flops
+    const SpatialVec& A_GP  = allA_GB[parent->getNodeNum()];
+    const SpatialVec  APlus = ~phi * A_GP;
```

Comment:
I'm finding this confusing -- A_GP and APlus aren't really accelerations here, right? They are `sqrt(M^-1)*f` rather than `M^-1*f` so wouldn't have the right units even if the input vector were a force.

Can you please make the documentation more precise? I think in general the method needs more documentation for each step so that a reader doesn't have to try to guess what you are doing. It also might help if you can point to a specific algorithm and page number in the literature if this code is a precise implementation of something there.

In general I think all the new code needs a quick pass through to take a skeptical look at the documentation to make sure there are no copy-and-paste errors leftover from the code this was derived from. Please take a look.